temporalio/client.py

```diff
@@ -0,0 +1,935 @@
+            SignalWorkflowInput(
+                id=self._id,
+                run_id=run_id,
+                signal=name,
+                args=args,
+            )
+        )
+
+    async def terminate(
+        self,
+        *args: Any,
+        reason: Optional[str] = None,
+        run_id: Optional[str] = SELF_RUN_ID,
+        first_execution_run_id: Optional[None],
+    ) -> None:
+        """Terminate the workflow.
+
+        Args:
+            args: Details to store on the termination.
+            reason: Reason for the termination.
+            run_id: Run ID to terminate. Defaults to using :py:meth:`run_id`. If
+                set to None or there is no :py:meth:`run_id`, this will
+                terminate the latest run for the workflow ID.
+            first_execution_run_id: First run ID that started the workflow. If
+                set, the termination makes sure that the workflow was started
+                with the given run ID.
+
+        TODO(cretz): Raises
+        """
+        if run_id == WorkflowHandle.SELF_RUN_ID:
+            run_id = self._run_id
+        await self._client._impl.terminate_workflow(
+            TerminateWorkflowInput(
+                id=self._id,
+                run_id=run_id,
+                args=args,
+                reason=reason,
+                first_execution_run_id=first_execution_run_id,
+            )
+        )
+
+
+@dataclass
+class StartWorkflowInput:
+    """Input for :py:meth:`OutboundInterceptor.start_workflow`."""
+
+    workflow: str
+    args: Iterable[Any]
+    id: str
+    task_queue: str
+    execution_timeout: Optional[timedelta]
+    run_timeout: Optional[timedelta]
+    task_timeout: Optional[timedelta]
+    id_reuse_policy: WorkflowIDReusePolicy
+    retry_policy: Optional[temporalio.common.RetryPolicy]
+    cron_schedule: str
+    memo: Optional[Mapping[str, Any]]
+    search_attributes: Optional[Mapping[str, Any]]
+    header: Optional[Mapping[str, Any]]
+    start_signal: Optional[str]
+    start_signal_args: Iterable[Any]
+
+
+@dataclass
+class CancelWorkflowInput:
+    """Input for :py:meth:`OutboundInterceptor.cancel_workflow`."""
+
+    id: str
+    run_id: Optional[str]
+    first_execution_run_id: Optional[str]
+
+
+@dataclass
+class QueryWorkflowInput:
+    """Input for :py:meth:`OutboundInterceptor.query_workflow`."""
+
+    id: str
+    run_id: Optional[str]
+    query: str
+    args: Iterable[Any]
+    reject_condition: Optional[WorkflowQueryRejectCondition]
+
+
+@dataclass
+class SignalWorkflowInput:
+    """Input for :py:meth:`OutboundInterceptor.signal_workflow`."""
+
+    id: str
+    run_id: Optional[str]
+    signal: str
+    args: Iterable[Any]
+
+
+@dataclass
+class TerminateWorkflowInput:
+    """Input for :py:meth:`OutboundInterceptor.terminate_workflow`."""
+
+    id: str
+    run_id: Optional[str]
+    first_execution_run_id: Optional[str]
+    args: Iterable[Any]
+    reason: Optional[str]
+
+
+class Interceptor:
```

Comment:
Why is this class needed?

Reply:
It just feels like a very OO old school approach.
In languages where functions are first class citizens this makes less sense to me.

Why not give users the freedom to choose the style they want:

- `Client(interceptors={"outbound": [ClientInterceptor1, ClientInterceptor2]}`
- `Worker(interceptors={"workflow": [lambda workflow_info: {"outbound": WorkflowOutbound(), "inbound": WorkflowInbound()})`

-- OR --

- `Client(interceptors=[my_interceptor.intercept_client])`
- `Worker(interceptors={'workflow': [my_interceptor.intercept_workflow]})`

Reply:
> In languages where functions are first class citizens this makes less sense to me.

FWIW, they are in Go and Java too and we chose interfaces there as well

> Why not give users the freedom to choose the style they want:

This is in stark contrast to your previous statements about reduction of user freedom.

Having said that, I can change this to accept `Iterable[Callable[temporalio.client.OutboundInterceptor, temporalio.client.OutboundInterceptor]]`.

The struggles here are: 1) the type is more complex for readers, 2) the documentation is a bit more difficult than on the class itself, 3) we cannot add an inbound interceptor or any other type, 4) it is hard to ship multi-purpose (i.e. client and worker) interceptors as single classes

#4 is pretty important. In Go I can basically `interceptors=[OpenTelemetryInterceptor]` and it will work throughout. Client interceptors that implement `WorkerInterceptor` are also used as worker interceptors (which works with activities and workflows).

Can you give your idea of how general client/workflow/activity interceptor from a helper library is configured on client and worker?

Reply:
> The struggles here are: 1) the type is more complex for readers, 2) the documentation is a bit more difficult than on the class itself, 3) we cannot add an inbound interceptor or any other type, 4) it is hard to ship multi-purpose (i.e. client and worker) interceptors as single classes

1. I agree, this is less readable to me too. I'm wondering if it's just me not being used to python typing syntax though because in TS the function signature type is easier for me to grasp.
2. You could create a TypedDict and put the docs there.
3. That can be done the way I described the workflow interceptor registration (returns a mapping of inbound and outbound interceptors).
4. I don't know if it's hard, just bundle them up in a single class and have an `get_worker_interceptors` and `get_client_interceptors` method, pretty much like I showed in the example above.

If you really like the class based approach I'm fine with it. I just want to make sure we're not making this harder to use than it has to be and that we have a pythonic API.

Reply:
For clarity, in Go you simply do a `opentracing.NewTracingInterceptor` in one line for client options and you've now enabled tracing for the client, workers, etc. It sounds much more confusing the TS way passing around functions and such.

I can take a class or callable to alleviate these concerns, but I think accepting a class really helps you share interceptor concerns instead of forcing people to make separate calls to get separate interceptors.

I was wanting to allow client-configured interceptors to work for workers too if they implemented `WorkerInterceptor`.

Reply:
After discussion, we will accept the union

Reply:
Done